src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -102,19 +102,21 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     	CacheService cacheService = new CacheService(dto);
     	MessageCacheItem cacheItem = cacheService.getCacheOfComponent();
     	if (cacheItem != null) { // Item is in cache
-    		if (cacheItem.getCachedData().isEmpty()) { // This means that the data to be used is from a fallback locale.
-				// If expired, try to first create and store cacheItem for the requested locale in a separate thread.
-				if (cacheItem.isExpired())
-					this.createCacheItemTask(null); // Pass null so that locale fallback will not be applied.
-
-    			// Use cached data of cacheItem.getLocale() --> fallback locale
-				MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), cacheItem.getLocale(), dto.getProductID(), dto.getVersion());
-				cacheItem = new ComponentService(fallbackLocaleDTO).getMessages(null);
-			} else if (cacheItem.isExpired()) {
+    		if (cacheItem.isExpired()) {
     			// Refresh the cacheItem in a separate thread
     			refreshCacheItemTask(cacheItem);
     		}
-    	} else { // Item is not in cache. Create and store cacheItem for the requested locale
+    	} else { // Item is not in cache.
+			ProductService ps = new ProductService(dto);
+			if (!ps.isSupportedLocale(Locale.forLanguageTag(dto.getLocale()))) { // Requested locale is not supported
+				Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new LinkedList<>(ps.getSupportedLocales()), Locale.forLanguageTag(dto.getLocale()));
+				if (ps.isSupportedLocale(matchedLocale)) { // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+					MessagesDTO matchedLocaleDTO = new MessagesDTO(dto.getComponent(), matchedLocale.toLanguageTag(), dto.getProductID(), dto.getVersion());
+					return new ComponentService(matchedLocaleDTO).getMessages(null);
```

Comment:
It's obscure to pass null to getMessages. Hard to understand.

Reply:
This has been there since before this PR, so I suggest you open another issue to address this specific concern. 

To understand the behavior when passing null to getMessages, read the javadoc of getMessages:

	 * @param fallbackLocalesIter The locale fallback queue to be used on failure. If null, there will be no fallback mechanism on failure so the message map will be empty.